### PR TITLE
Application offers and connections

### PR DESCRIPTION
**Pull request by SimonRichardson**
### Checklist

 - [-] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [-] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [-] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

This adds offers and offer connections to the importing and exporting migration steps.

## QA steps

To follow!

### CODE REVIEW

File: state/migration_export.go [hunk 1]
```diff
@@ -1226,6 +1233,42 @@ func (e *exporter) remoteEntities() error {
...
 }
 
+// offerConnectionsShim provides a way to model our dependencies by providing
+// a shim layer to manage the covariance of the state package to the migration
+// package.
+type offerConnectionsShim struct {
+	st *State
+}
+
+func (s offerConnectionsShim) AllOfferConnections() ([]migrations.MigrationOfferConnection, error) {
```
Comment (manadart) [inline state/migration_export.go]: Needs comment.

Comment (manadart) [inline state/migration_export.go]: Needs comment.

File: state/migrations/doc.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package migrations aims to create an intermidiate state between state and the
+// description package. One that aims to correctly model the dependencies
+// required for migration.
+//
+// This package is in a state of transition, from the traditional way in the
```
Comment (manadart) [inline state/migrations/doc.go]: Nice.

File: state/userpermission.go [hunk 3]
```diff
@@ -114,16 +114,20 @@ func removePermissionOp(objectGlobalKey, subjectGlobalKey string) txn.Op {
...
+	doc := makePermissionDoc(objectGlobalKey, subjectGlobalKey, access)
 	return txn.Op{
 		C:      permissionsC,
 		Id:     permissionID(objectGlobalKey, subjectGlobalKey),
 		Assert: txn.DocMissing,
 		Insert: doc,
 	}
 }
+
+func makePermissionDoc(objectGlobalKey, subjectGlobalKey string, access permission.Access) *permissionDoc {
```
Comment (manadart) [inline state/userpermission.go]: Is `newPermissionDoc` a better name for this due to returning a reference?

File: state/migration_export.go [hunk 4]
```diff
@@ -67,6 +67,7 @@ type ExportConfig struct {
 	SkipRelationData         bool
 	SkipInstanceData         bool
 	SkipApplicationOffers    bool
+	SkipOfferConnections     bool
```
Comment (manadart) [inline state/migration_export.go]: We should be setting these to true for bundle exports when we add new ones that it doesn't need.

File: state/migration_import.go [hunk 5]
```diff
@@ -1607,8 +1752,8 @@ func (i *importer) remoteEntities() error {
 	return nil
 }
 
-//go:generate mockgen -package state -destination migration_import_mock_test.go github.com/juju/juju/state TransactionRunner,StateDocumentFactory,DocModelNamespace,RemoteEntitiesDescription,RelationNetworksDescription,RemoteApplicationsDescription
-//go:generate mockgen -package state -destination migration_description_mock_test.go github.com/juju/description RemoteEntity,RelationNetwork,RemoteApplication,RemoteSpace,Status
+//go:generate mockgen -package state -destination migration_import_mock_test.go github.com/juju/juju/state TransactionRunner,StateDocumentFactory,DocModelNamespace,RemoteEntitiesDescription,RelationNetworksDescription,RemoteApplicationsDescription,ApplicationOfferStateDocumentFactory,ApplicationOfferDescription
```
Comment (manadart) [inline state/migration_import.go]: Nit: I'd be inclined to move this higher up, at least above the first declaration of something that we're mocking.

File: state/userpermission.go (hunk 3 above)
Comment (SimonRichardson) [reply state/userpermission.go]: Yes, yes it would be! Good catch!

File: state/migration_export.go (hunk 4 above)
Comment (SimonRichardson) [reply state/migration_export.go]: Ah, I hadn't thought about that!

File: state/migration_import.go (hunk 5 above)
Comment (SimonRichardson) [reply state/migration_import.go]: I'm a fan of moving these to `package_test.go` - thoughts on that?

Comment (manadart) [reply state/migration_import.go]: Works for me.

**Comment by SimonRichardson** [comment]
https://trello.com/c/3rIc0Vd7/2071-migrate-applicationoffers-applicationofferconnections

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**